### PR TITLE
Extension build configuration refactor

### PR DESCRIPTION
**Pull request by samansmink**
This PR introduces a bit of a refactor to the way extensions are configured when building DuckDB.

# Goals
- provide a flexible interface for devs to specify which extensions should be built
- allow mixing in-tree-extension and out-of-tree extensions from various sources in builds
- use the same mechanism to build in-tree extensions and out-of-tree extensions
- centralize configuration of which extensions to build and when.

# How to build extensions
For a more detailed description, checkout `extension/README.md` which I added in this PR.
To summarize, there are now various ways to specify which extensions should be built. Some examples:

## From command line
To build just an in-tree extension such as icu, the old makefile variables still work
```sh
BUILD_ICU=1 make
```

A new addition is the DUCKDB_EXTENSIONS variable which also allows passing out-of-tree extensions (which should manually be placed in './extensions_external') For example, to build duckdb, with the in-tree json and icu extensions, and also the out-of-tree arrow extension:
```sh
DUCKDB_EXTENSIONS='json;icu;arrow' make
```

## From config file
A new addition is the ability to specify which extensions to build in a config file. For example to build the postgres scanner and also the icu extension the file `extension_config.cmake` containing:
```cmake
duckdb_extension_load(icu)
duckdb_extension_load(postgres_scanner
    DONT_LINK
    GIT_URL https://github.com/duckdblabs/postgres_scanner
    GIT_TAG cd043b49cdc9e0d3752535b8333c9433e1007a48
)
```
now to build using this configuration, run:
```sh
EXTENSION_CONFIGS='extension_config.cmake' make
```

### Default configuration files
There are a few places that now contain cmake config files to specify which extensions to build. Firstly, in `extension/extension_config.cmake` we specify which extensions are always built. This currently contains the parquet extension and the jemalloc extension. Additionally the `extension/extension_config_local.cmake` is gitignored and also checked by the build, and can be created by devs to specify which extensions they want built in an ad-hoc way.  I think this is a bit nicer than the previous approach of setting BUILD_<extname> in your bashrc. For more details, check out `extension/README.md`

# Nicer cmake logs
We now log on each build which extensions are built and which of those are actually linked in the binaries produced, for example heres the output from a build combining extensions from various sources:
```
...
-- Building extension 'json' from '/Users/sam/Development/duckdb/extensions'
-- Building extension 'substrait' from https://github.com/duckdblabs/substrait @ 48d64b27c7a6985d6f6c3e65044038544608c03b
-- Building extension 'parquet' from '/Users/sam/Development/duckdb/extensions'
-- Extensions linked into DuckDB: json, parquet
-- Extensions built but not linked: substrait
...
```

# TODOs remaining
## `package_build.py` builds (R, Node, Python)
For R, Node, Python and JDBC, i added an extension configuration file to specify which extensions are build. However this only works for cmake builds. I still want to unify the configuration for source
distributions because now there are still two files where we specify extensions and additionally it makes the code in `src/main/extension/extension_helper.cpp` a lot more ugly.

## Clean up CI extension config
Currenty there are still a few jobs which specify some custom combination of extensions (now using the `BUILD_EXTENSIONS=` cmake var). I think we want to minimize/unify this as much as possible to
keep the number of configurations there to a minimum

## Misc
I will propagate these changes to extension-template & other out-of-tree extensions

### CODE REVIEW

**Comment by carlopi** [comment]
Looks great to me!

As discussed with @samansmink already, WebAssembly.yml was bound to fail, it can be potentially disabled entirely and I can then restore support adapting WebAssembly CI job after this rework lands.

**Comment by Mytherin** [comment]
Thanks for the PR! This looks great!

I am somewhat worried about merge conflicts between feature/master as we will be continuing the feature branch for a while - perhaps you could split this up and PR a part of the more benign changes to the master branch - e.g. the renames of all of the extensions and moving of extension headers, as well as the more benign Makefile changes (renaming `EXTENSIONS` to `CMAKE_VARS`)?

**Comment by samansmink** [comment]
> I am somewhat worried about merge conflicts between feature/master as we will be continuing the feature branch for a while - perhaps you could split this up and PR a part of the more benign changes to the master branch - e.g. the renames of all of the extension C++ classes and moving of extension headers, as well as the more benign Makefile changes (renaming `EXTENSIONS` to `CMAKE_VARS`)?

Yea good point, wasn't really sure how to go forward there as this pr is a bit of a walking merge conflict, but that sounds like a good solution.

So i'll leave this PR open, apply the less scary, but high merge-conflictable changes in a separate pr to master. Then when master gets merged into feature, i merge this pr with feature again

File: .github/workflows/ExtensionRebuild.yml [hunk 1]
```diff
@@ -202,15 +189,7 @@ jobs:
...
-        visualizer: 0
-        icu: 0
-        tpch: 0
-        tpcds: 0
-        httpfs: 0
-        fts: 0
-        json: 0
-        excel: 0
-        sqlsmith: 0
+        build_out_of_tree_extensions: 0
```
[the remark this answers is not in the document]
Comment (samansmink) [reply .github/workflows/ExtensionRebuild.yml]: good catch, fixed it!

File: scripts/generate_extensions_function.py [hunk 2]
```diff
@@ -26,12 +26,23 @@
...
+def parse_extension_configs():
+    ext_configs = [
+        os.path.join('..', '.github', 'config', 'out_of_tree_extensions.cmake'),
+        os.path.join('..', '.github', 'config', 'in_tree_extensions.cmake')
+    ]
+    extension_names = []
+    for file in ext_configs:
+        with open(file, 'r') as file:
+            data = file.read().replace('\n', '')
+        pattern = re.compile(r"duckdb_extension_load\s*\(\s*([A-z\_0-9]*)\s")
```
Comment (carlopi) [inline scripts/generate_extensions_function.py:40]: Somehow a nit, but this seems to be only a regex on the cmake file, so stuff like `if (! WIN32)` has no effect (unsure whether this is already taken into account).

**Review by carlopi — APPROVED** [review]

File: .github/config/out_of_tree_extensions.cmake [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    GIT_URL https://github.com/duckdblabs/postgres_scanner
+    GIT_TAG cd043b49cdc9e0d3752535b8333c9433e1007a48
+)
+duckdb_extension_load(spatial
+    DONT_LINK
+    GIT_URL https://github.com/duckdblabs/duckdb_spatial.git
+    GIT_TAG f577b9441793f9170403e489f5d3587e023a945f
+)
+
+if (NOT WIN32)
```
[the remark this answers is not in the document]
Comment (samansmink) [reply .github/config/out_of_tree_extensions.cmake:24]: I think changing this config file in the future is perfectly fine, i just went with matching the previous behaviour.

With regards to "having the source of truth in the extension itself", i'm not sure if i agree (or fully understand?). Some extensions may not build on the DuckDB windows job, but would run on the extension's own CI. Therefore i would say you want to be able to control from duckdb which extensions are built here in the config.

File: .github/config/out_of_tree_extensions.cmake [hunk 4]
```diff
@@ -0,0 +1,35 @@
+#
+# This is the DuckDB out-of-tree extension config as it will run on the CI
+#
+# to build duckdb with this configuration run:
+#   EXTENSION_CONFIGS=.github/config/extension_config.cmake make
```
[the remark this answers is not in the document]
Comment (samansmink) [reply .github/config/out_of_tree_extensions.cmake:5]: oops good catch, should be renamed here indeed

File: .github/config/out_of_tree_extensions.cmake (hunk 3 above)
[the remark this answers is not in the document]
Comment (carlopi) [reply .github/config/out_of_tree_extensions.cmake:24]: I agree it's not incompatible / can be done in the future.
And "source of truth" I mean that the answer "is extension X compatible with Windows?" is in my mind the extension business, while "I would like extension X" is config business, but they are somehow intertwined (but that's also inevitable to some extent).

I'd say good / resolved.

File: scripts/generate_extensions_function.py (hunk 2 above)
Comment (samansmink) [reply scripts/generate_extensions_function.py:40]: yea we may want to look into doing this in some other way. I would vote to keep this slightly hacky solution for now, then look into improving it as we need

**Comment by Mytherin** [comment]
Thanks! LGTM

**Pull request merged**